pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -1556,17 +1539,95 @@ class Timestamp(_Timestamp):
 
-        _date_attributes = [year, month, day, hour, minute, second,
-                            microsecond, nanosecond]
+        args_len = len(args)
+
+        # GH 30543 if pd.Timestamp already passed, return it
+        # check that only ts_input is passed
+        # checking verbosely, because cython doesn't optimize
+        # list comprehensions (as of cython 0.29.x)
+        if args_len == 1 and len(kwargs) == 0 and isinstance(args[0], _Timestamp):
+            return args[0]
+
+        # Building from ts_input
+        if args_len == 1:
+            if kwargs:
+                if ("year" in kwargs or "month" in kwargs or "day" in kwargs or
+                        "hour" in kwargs or "minute" in kwargs or "second" in kwargs or
+                        "microsecond" in kwargs):
+                    raise ValueError("Cannot pass a date attribute keyword argument")
+                if isinstance(args[0], str):
+                    if "nanosecond" in kwargs:
+                        raise ValueError(
+                            "Cannot pass a date attribute keyword "
+                            "argument when passing a date string; 'tz' is keyword-only"
+                        )
+                    if any(k not in ["tz", "tzinfo", "unit"] for k in kwargs.keys()):
+                        raise ValueError(
+                            "When passing a date string "
+                            "can only pass unit and tz or tzinfo as a keyword argument."
+                        )
+
+            ts_input = args[0]
+            tzinfo = kwargs.get("tzinfo")
+        # Building from positional arguments
+        elif 9 > args_len > 2 and isinstance(args[1], int):
+            args = args + (None,) * (8 - args_len)
+            year, month, day, hour, minute, second, microsecond, tzinfo = args
+
+            if kwargs:
+                # Positional or keyword arguments
+                err_msg = (
+                    "argument for function given by name ('{}') and position ({})"
+                )
+
+                datetime_components = ["year", "month", "day", "hour", "minute",
+                                       "second", "microsecond", "tzinfo"]
+                for i, key in enumerate(datetime_components):
+                    if args_len > i:
+                        if key in kwargs:
+                            raise TypeError(err_msg.format(key, i))
+                    else:
+                        break
+
+                hour = kwargs.get("hour", hour)
+                minute = kwargs.get("minute", minute)
+                second = kwargs.get("second", second)
+                microsecond = kwargs.get("microsecond", microsecond)
+                tzinfo = kwargs.get("tzinfo", tzinfo)
+        # Keywords only
+        elif args_len == 0:
+            ts_input = kwargs.get("ts_input", _no_input)
+            year = kwargs.get("year")
+            month = kwargs.get("month")
+            day = kwargs.get("day")
```

Comment:
is there anyway to simply translate the args to kwargs near the top

this looks so fragile and error prone